### PR TITLE
Expand SkyCoord.guess_from_table

### PR DESCRIPTION
**Pull request by nstarman**
### Description

1) Allow for multiple regex matches when they are exact component names.
   eg: allow for column “dec” and “pm_dec” to work
2) allow for differential component names
  eg:  "pm_ra_cosdec", “pm_dec”
3) improve error message to show all matches, if there are more than one.

- [x] write tests
- [x] changelog
- [x] docstring
- [ ] docs (?)

Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### CODE REVIEW

File: astropy/coordinates/sky_coordinate.py
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py:1926]: I know this was in the original version, but maybe this should be a smidge clearer here, like:
```suggestion
                       ': "{1}". Cannot guess coordinates from a table with this ambiguity.')
```
or similar?

**Review by eteq — APPROVED** [review]
Thanks @nstarman - My initial thought on seeing this was "feature creep!" since this method is supposed to be limited in scope. but now that I look at it I think this is good because it really is very logical and unambiguous!  Just one minor wording suggestion.

**Comment by nstarman** [comment]
> "feature creep!" since this method is supposed to be limited in scope.

@eteq, I refrained from further modifications, but would be interested in pushing more if it's not feature creep. In particular:

- checking for parallax / redshift if distance isn't found and using ``coord.Distance()``
- checking for exact component name matches to break ambiguity; e.g. "ra" is preferred over "RA[J2000]" if both are in the table. This might be a flag "prefer_exact_match", or something.

Do you think any of these would make for another good PR?

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:1889]: Nitpick, but it seems likely that it is faster to just do
```
= tuple(frame.get_representation_component_names()) + tuple(frame.get_representation_component_names("s"))
```

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:1903]: indent off?

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:1917]: I think I'd use a `set` for matches, so one can do
```
matches -= representation_component_names - {comp_name}
```

Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py]: I like the usage of sets. I've changed it over so that all comparison operations are set-based.

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py]: yes, thanks. I fixed it by all making the previous argument shorter using f-strings.

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py]: agreed. I'm changing over all relevant lists to sets.

**Comment by nstarman** [comment]
@mhvk, @eteq, @adrn. Tests passing (3.7 failure is unrelated).

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:1881]: "for value" -> "for a value"

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:1923]: I'd just put an f-string inside `ValueError`, but am saying this only because of the typo above means another commit may be necessary...

File: astropy/coordinates/tests/test_sky_coord.py
Comment (mhvk) [inline astropy/coordinates/tests/test_sky_coord.py:1290]: Oops... nice catch

File: astropy/coordinates/tests/test_sky_coord.py
Comment (mhvk) [inline astropy/coordinates/tests/test_sky_coord.py:1304]: For clarity, I'd just do
```
tab['pm_ra_cosdec'] = np.random.rand(1000)*u.mas/u.yr
```

Aside: super-nitpick and not your problem, but would you might changing the size to `10` (also above)? Might as well focus on testing what we're interested in rather than generating excess random numbers...

File: astropy/coordinates/tests/test_sky_coord.py
Comment (mhvk) [inline astropy/coordinates/tests/test_sky_coord.py:1317]: This can be `with pytest.raises(u.UnitTypeError, match="no unit was given.")` (note the more exact exception type)

**Review by mhvk — APPROVED** [review]
Only nitpicks left. Since they are really nitpicks, I'll approve. Nice!

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1881]: ✅

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1923]: ✅

File: astropy/coordinates/tests/test_sky_coord.py
Comment (nstarman) [reply astropy/coordinates/tests/test_sky_coord.py:1304]: ✅

File: astropy/coordinates/tests/test_sky_coord.py
Comment (nstarman) [reply astropy/coordinates/tests/test_sky_coord.py:1317]: ✅ . Nice. I didn't know pytest.raises had that filter built in.

File: astropy/coordinates/tests/test_sky_coord.py
Comment (nstarman) [reply astropy/coordinates/tests/test_sky_coord.py:1290]: ✅

**Comment by mhvk** [comment]
Super! This is a really nice addition and nicely implemented too!

Now the tests have passed, and I'd love to press the merge button, but for some reason it tells me it can only be done with 6 approving reviews!?!?!?!?  Anybody any idea what's going on?

**Comment by pllim** [comment]
Don't merge yet. We're freezing merges until `towncrier` is in place.

**Comment by nstarman** [comment]
@adrn @mhvk @eteq. Rebased to fix conflict. Test failures are unrelated. Switched change-log to Towncrier news fragment.

**Comment by adrn** [comment]
I think the test failures are unrelated (see #11458) but @pllim could confirm?

**Comment by pllim** [comment]
What timing. matplotlib dev just got fixed this morning, so everything should be green now except mpldev job on CircleCI. Might need a rebase to pick up RTD fix that @nstarman himself put in. 😄

**Comment by adrn** [comment]
Looks like the RTD fix is in! So, the mpldev failure is an expected failure, and I will merge. Thanks @nstarman !

**Pull request merged**
